man/schedctl.8
.TH "schedctl" 8
.SH "NAME"
schedctl

.SH "SYNOPSIS"
.B schedctl [--stop | --start | --status | --reread-policy | --savestate]

.B schedctl [--score | --inherit] jobid1 .. jobidN

.SH "DESCRIPTION"
.TP
Commands to control job scheduling.

.SH "OPTIONS"
.TP
.B \-d \-\-debug
Turn on communication debugging.
.TP
.B \-h \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-\-version
Displays client revision and Cobalt version
.TP
.B \-\-stop
stop scheduling jobs
.TP
.B \-\-start
resume scheduling jobs
.TP
.B \-\-status
query scheduling status
.TP
.B \-\-reread-policy
reread the utility function definition file
.TP
.B \-\-savestate
write the current state to the specified file
.TP
.B \-\-score
adjust the scores of the jobid arguments
.TP
.B \-\-inherit
control the fraction of the score inherited by jobs which depend on the jobid arguments

.SH "SEE ALSO"

.SH "BUGS"


CHANGES
Changes from previous Cobalt Versions

Changes for 0.99.26
- Corrected a bug that was causing messages in the clients that used to go to
  stdout to go to stderr instead

Changes for 0.99.25
- Fixes for clients that were causing erronious default values for proccount
  have been fixed.
- The setgid wrapper now calls a site-selectable python interperter with the
  '-E' flag by default.
- A bug that could cause TimeRemaining to show negative values has been fixed
- Queues may be added or removed from blocks without overwriting the entire
  list with new options in partadm using --rmq or --appq with the --queue option.


Changes for 0.99.24
* Reverting client changes.  Found some issues when deploying to test systems
* --defer flag has been reverted with these as a casualty.

Changes for 0.99.23
* Client code has undergone a refactor.  Numerous bugs have been addressed
  with these that were found while improving test coverage of this code
* IO Block booting and automatic control has been added.  Please consult the
  partadm manpage for details
* This release requires pybgsched-0.1.3
* A user may voluntarily set their job's score to 0 and let it re-accrue using
  qalter --defer (Ticket #590)

Changes to 0.99.22
* Fix for a deadlock that could occur due to a race condition in the 
  boot-reaping code.
* Fix for a pid collision that could occur after a forced delete of a job in
  the system-script-forker component leading to a job getting stuck in the
  Job_Epilogue state.
* python 2.7 and it's change to the XMLRPC libraries are now supported.

Changes to 0.99.21
* Minor fix release to correct a failure mode where we can lose a process group
  if a job's boot is delayed.

Changes to 0.99.20
* Small changeset to fix a race condition that was hanging jobs that came in
   on 0.99.19
* Better error logging for situations where the boots and the process group
  information haven't lined up yet.  This usually happens during high-load and
  can result in block boot completion never being detected.

Changes to 0.99.19
* Manual boot control has been added
* major refactor of compute block rebooting

Changes to 0.99.18
* Fix for other systems that were seeing erronious geometry data
* SoftwareFailure now accurately reported

Changes to 0.99.17

* Support for --geometry has been added.  This will constrain a job to run 
blocks having a specific geometry in nodes, partadm, partlist and qstat have
been updated to show this.
* Support for passthrough blocking reservations has been added.
* Fixed a bug where a job run outside of Cobalt would not properly block resources.
* Removed autoreboot in the event of a block entering a control system 
  SoftwareFalure state on BG/Q.  Normal cleanup should handle this now.

Changes to 0.99.15
* Fix for an issue that was causing partadm to malfunction when reservations
were set on the system.

Changes to 0.99.12
* As a note, this is a release of fixes and backported features from BG/Q
* Fixed issues that could cause a bad hardware state to be overridden as 
"blocked" allowing the block to be drained
* Fix for a situation where a partition that is inelligible to run due to a
pending reservation is selected for draining.

Changes to 0.99.11
* Fix for an additional circumstance that can cause a small reservation to
collapse a backfill window to zero.

Changes to 0.99.10
* Fixed an issue that was introduced in 0.99.6 where children were not being
properly added to a block.  This caused strange behavior in reservations,
including failure to run jobs smaller than the reservation or a small job
running on the largest block in a reservation
* Refactored the update_block_state code to operate more similarly to the way
BG/P is currently doing it.  This should correct an issue where a block can
have an error state overridden.
* Cleaned up a circumstance where wires between passthrough midplanes can be
added to a block's wire list.  This can cause erronious wiring conflicts to be
detected.
* Fixed an issue where duplicate wires were being added

Changes to 0.99.9

* Fixed an issue where a job could choose a block for draining that is blocked
by a pending reservation.  This would cause the backfill time to collapse to zero.
* Added code to take blocks offline if passthrough nodeboards reported their state
as in error.  It does not offline them if it is only a compute node down, it must
be the nodeboard itself in error as would be set by a BOARD_IN_ERROR control action.

Changes to 0.99.8

* Fixed an issue where dimensions without cables (like the A,B, and C
  dimensions on a 2 midplane system) would cause the system component to fail
  on boot.

Changes to 0.99.7

* Cable errors are now properly detected
* IO Node errors in available links are detected
* Due to changes, libpybgsched v 0.1.1 is required.  This also
  means the V1R1M1 driver changes can also be enabled.

Changes to 0.99.6

* Performance improvements over 0.99.5 in update_block_state and wiring 
  conflict detection.  This makes the partadm commands far more responsive.
* Corrected an error in a block's children that affected cold (non-statefile)
  startup.

Changes to 0.99.0pre36

* Fixed bug in resource reservation that would leave a reserved partition in
  the 'idle' state
* Updated the code that sets the kernel profile for a partition to set the
  profile even if no profile is currently set


Changes to 0.99.0pre35

* Fixed bug where pickling IncrID could fail when deleting db related
  attributes from the object's dictionary


Changes to 0.99.0pre34

* Time spent in the hold state is now exported to the utility function
* bgsystem now tracks forked tasks using a (forker, id) tuple instead of just
  id in order to prevent id collisions in the active jobs list
* Updated cleanup and resource reservation in bgsystem updated to prevent
  multiple partition cleanups for a single job
* Forker classes updated so that logging messages properly identify the
  component from which they came

src/clients/schedctl.py
#!/usr/bin/env python
"""
Commands to control job scheduling.

Usage: %prog [--stop | --start | --status | --reread-policy | --savestate]
Usage: %prog [--score | --inherit] jobid1 .. jobidN

version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS:

'-d','--debug',dest='debug',help='turn on communication debugging',callback=cb_debug
'--stop',action='store_true',dest='stop',help='stop scheduling jobs'
'--start',action='store_true',dest='start',help='resume scheduling jobs'
'--status',action='store_true',dest='stat',help='query scheduling status'
'--reread-policy',action='store_true',dest='reread',help='reread the utility function definition file'
'--savestate',dest='savestate',type='string',help='write the current state to the specified file',callback=cb_path
'--score',dest='adjust',type='string',help='<jobid> <jobid> adjust the scores of the arguments',callback=cb_score

'--inherit',dest='dep_frac',type='float', /
  help='<jobid> <jobid> control the fraction of the score inherited by jobs which depend on the arguments'

"""
import logging
import sys

from Cobalt import client_utils
from Cobalt.client_utils import cb_debug, cb_path, cb_score

from Cobalt.arg_parser import ArgParse

__revision__ = 'TBD'
__version__  = 'TBD'

SCHMGR = client_utils.SCHMGR
QUEMGR = client_utils.QUEMGR

def validate_args(parser, args):
    """
    Validate arguments
    """
    spec     = {} 
    opts     = {} 
    opt2spec = {}
    
    opt_count = client_utils.get_options(spec, opts, opt2spec, parser)

    if opt_count == 0:
        parser.parser.print_help()
        client_utils.logger.error("Need at least one option")
        sys.exit(1)

    # Make sure jobid or queue is supplied for the appropriate commands
    args_opts =  [opt for opt in spec if opt in ['adjust','dep_frac']]
    if parser.no_args() and args_opts:
        client_utils.logger.error("At least one jobid must be supplied")
        sys.exit(1)

    elif not parser.no_args():
        if not args_opts:
            client_utils.logger.error("No arguments needed")
        else:
            for i in range(len(args)):
                if args[i] == '*':
                    continue
                try:
                    args[i] = int(args[i])
                except ValueError:
                    client_utils.logger.error("jobid must be an integer, found '%s'" % args[i])
                    sys.exit(1)

    # Check mutually exclusive options
    mutually_exclusive_option_lists = [['stop', 'start', 'stat', 'reread', 'savestate', 'adjust'],
                                       ['stop', 'start', 'stat', 'reread', 'savestate', 'dep_frac']]

    if opt_count > 1:
        client_utils.validate_conflicting_options(parser, mutually_exclusive_option_lists)

def main():
    """
    schedctl main function.
    """
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    use_cwd = False
    options = {}

    # list of callback with its arguments
    callbacks = [
        # <cb function>     <cb args (tuple) >
        [ cb_debug        , () ],
        [ cb_score        , () ],
        [ cb_path         , (options, use_cwd) ] ]

    # Get the version information
    opt_def =  __doc__.replace('__revision__', __revision__)
    opt_def =  opt_def.replace('__version__', __version__)

    parser = ArgParse(opt_def, callbacks)

    # Set required default values: None

    parser.parse_it() # parse the command line
    args  = parser.args
    opt   = parser.options

    whoami = client_utils.getuid()

    validate_args(parser, args)

    if opt.stop != None:
        client_utils.component_call(SCHMGR, False, 'disable', (whoami,))
        client_utils.logger.info("Job Scheduling: DISABLED")
        sys.exit(0)
    elif opt.start != None:
        client_utils.component_call(SCHMGR, False, 'enable', (whoami,))
        client_utils.logger.info("Job Scheduling: ENABLED")
        sys.exit(0)
    elif opt.stat != None:
        if client_utils.component_call(SCHMGR, False, 'sched_status', ()):
            client_utils.logger.info("Job Scheduling: ENABLED")
        else:
            client_utils.logger.info("Job Scheduling: DISABLED")
        sys.exit(0)
    elif opt.reread != None:
        client_utils.logger.info("Attempting to reread utility functions.")
        client_utils.component_call(QUEMGR, False, 'define_user_utility_functions', (whoami,))
        sys.exit(0)
    elif opt.savestate != None:
        response = client_utils.component_call(SCHMGR, False, 'save', (opt.savestate,))
        client_utils.logger.info(response)
        sys.exit(0)

    if opt.adjust != None:
        client_utils.set_scores(opt.adjust, args, whoami)

    if opt.dep_frac != None:
        specs = [{'jobid':jobid} for jobid in args]
        response = client_utils.component_call(QUEMGR, False, 'set_jobs', (specs, {"dep_frac": opt.dep_frac}, whoami))

        if not response:
            client_utils.logger.info("no jobs matched")
        else:
            dumb = [str(r["jobid"]) for r in response]
            client_utils.logger.info("updating inheritance fraction for jobs: %s" % ", ".join(dumb))

if __name__ == '__main__':
    try:
        main()
    except SystemExit:
        raise
    except Exception, e:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", e)
        sys.exit(1)
